### PR TITLE
feat: provide openapi doc

### PR DESCRIPTION
**Pull request by n063h**
ref to #359 
# Description
Provide openapi json doc for users to build their api-reference or swagger site.

# How it works
mosec will inspect input/return variable to figure out their type schema in the python end, save them into json schema file. Only the input variable of Ingress and the return variable of Egress worker will be inspected.

Since Rust side can't have access to python varible type, so here I set placeholder types `InferenceRequest` and `InferenceResponse`, and replace them with the json schema generated in python side.

### CODE REVIEW

File: src/apidoc.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    fn merge_request(&self, req_body: &mut RequestBody, req_schema: RefOr<Schema>) {
+        let content = ContentBuilder::new().schema(req_schema).build();
+        req_body
+            .content
+            .insert("application/json".to_string(), content);
+    }
+
+    fn merge_response(&self, response: &mut Responses, res_schema: RefOr<Schema>) {
+        let ok_res = ResponseBuilder::new()
+            .content("application/json", Content::new(res_schema))
```
Comment (kemingy) [inline src/apidoc.rs:90]: Always JSON?

File: src/apidoc.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        &self,
+        api: &'a mut OpenApi,
+        route: &str,
+        method: &PathItemType,
+    ) -> Option<&'a mut RequestBody> {
+        let op = self.get_operation(api, route, method).unwrap();
+        op.request_body.as_mut()
+    }
+
+    fn get_route_responses<'a>(
```
Comment (kemingy) [inline src/apidoc.rs:63]: Is it necessary to have an explicit lifetime?

File: mosec/mixin/typed_worker.py [hunk 3]
```diff
@@ -27,32 +27,54 @@
...
+        params = list(sig.parameters.values())
+        if len(params) < 1:
+            raise TypeError("`forward` method doesn't have enough(1) parameters")
+        typ = params[index].annotation
+    else:
+        typ = sig.return_annotation
+        if typ is inspect.Signature.empty:
+            raise TypeError("`forward` method doesn't have return annotation")
+        if not inspect.isclass(typ):
+            typ = msgspec.inspect.type_info(typ).__class__
```
[the remark this answers is not in the document]
Comment (n063h) [reply mosec/mixin/typed_worker.py:47]: It's for the types which are not (primitive or defined by msgspec). 
For example, if the function returns Any, then sig.return_annotation doen't return a class, it returns an instance typing.Any. 

```python
class Preprocess(Worker):
    def forward(self, data: str)->Any:
        return [data*2]
typ=inspect.signature(Preprocess.forward).return_annotation
typ2=msgspec.inspect.type_info(typ).__class__
print(typ,typ2) # typing.Any <class 'msgspec.inspect.AnyType'>
print(json.schema(typ))  # {} ，can't expose any information
print(json.schema(typ2) # {'$ref': '#/$defs/AnyType', '$defs': {'AnyType': {'title': 'AnyType', 'description': 'A type corresponding to `typing.Any`.', 'type': 'object', 'properties': {}, 'required': []}}}
```

File: mosec/server.py [hunk 4]
```diff
@@ -363,6 +373,34 @@ def append_worker(
...
+        schema = {
+            "req_schema": req_schema,
+            "res_schema": res_schema,
+            "schemas": schemas,
+        }
+        tmp_path = pathlib.Path(os.path.join(self._configs["path"], MOSEC_OPENAPI_PATH))
+        tmp_path.parent.mkdir(parents=True, exist_ok=True)
+        tmp_path.write_text(
+            msgspec.json.encode(schema).decode("utf-8"), encoding="utf-8"
+        )
```
[the remark this answers is not in the document]
Comment (n063h) [reply mosec/server.py:402]: write_text requires a str param, I didn't try if bytes works

File: src/apidoc.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+use utoipa::openapi::{
+    path::Operation, request_body::RequestBody, Components, Content, ContentBuilder, OpenApi,
+    PathItemType, RefOr, ResponseBuilder, Responses, Schema,
+};
+
+#[derive(Deserialize, Default)]
+pub(crate) struct InferenceSchemas {
+    req_schema: RefOr<Schema>,
+    res_schema: RefOr<Schema>,
+    schemas: BTreeMap<String, RefOr<Schema>>,
```
[the remark this answers is not in the document]
Comment (n063h) [reply src/apidoc.rs:28]: just follow the lib

File: src/apidoc.rs [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+use serde::Deserialize;
+use utoipa::openapi::{
+    path::Operation, request_body::RequestBody, Components, Content, ContentBuilder, OpenApi,
+    PathItemType, RefOr, ResponseBuilder, Responses, Schema,
+};
+
+#[derive(Deserialize, Default)]
+pub(crate) struct InferenceSchemas {
+    req_schema: RefOr<Schema>,
+    res_schema: RefOr<Schema>,
```
[the remark this answers is not in the document]
Comment (n063h) [reply src/apidoc.rs:27]: ok

File: mosec/server.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kemingy) [reply mosec/server.py:402]: There is a [write_bytes](https://docs.python.org/3/library/pathlib.html#pathlib.Path.write_bytes) method.

File: mosec/mixin/typed_worker.py [hunk 7]
```diff
@@ -27,32 +27,54 @@
...
-        raise TypeError("`forward` method doesn't have enough(1) parameters")
+    if target == "parameters":
+        params = list(sig.parameters.values())
+        if len(params) < 1:
+            raise TypeError("`forward` method doesn't have enough(1) parameters")
+        typ = params[index].annotation
+    else:
+        typ = sig.return_annotation
+        if typ is inspect.Signature.empty:
+            raise TypeError("`forward` method doesn't have return annotation")
```
Comment (kemingy) [inline mosec/mixin/typed_worker.py:45]: Why would an empty response annotation raise an exception?

File: src/apidoc.rs (hunk 1 above)
Comment (n063h) [reply src/apidoc.rs:90]: yes, this is the point I was trying to ask. I remember having seen an "application/msgpack" somewhere, but have no idea how to test on it.

File: mosec/mixin/typed_worker.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kemingy) [reply mosec/mixin/typed_worker.py:47]: I guess we don't really need this. Even the `msgspec.insepct.AnyType` schema has no information.

Or if there are any other conditions that matter?

File: src/apidoc.rs (hunk 1 above)
Comment (kemingy) [reply src/apidoc.rs:90]: The return type can be get from the `--mime`. But I think you can pass the req & resp mime type through the same file.

File: src/apidoc.rs (hunk 2 above)
Comment (n063h) [reply src/apidoc.rs:63]: Actually I annotated lifetime on more functions which ware removed under clippy's warning. 
I didn't really understand this part, will continue to learn.

File: mosec/mixin/typed_worker.py (hunk 7 above)
Comment (n063h) [reply mosec/mixin/typed_worker.py:45]: I want to inform users you can and shall annotate the type to enable this feature, at least a Any type. It's also ok if err not raised.

File: mosec/mixin/typed_worker.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (n063h) [reply mosec/mixin/typed_worker.py:47]: No other notable conditions. If no information is ok, I will remove the class wrapper.

**Comment by kemingy** [comment]
Is it ready to review?

**Comment by n063h** [comment]
> Is it ready to review?

It works, but needs to be improved.
In the last commit, I tried to integrate the input/output/components schemas  to simplify the interface. But it led to extra components schema. I'm working on it to make it simple and correct.
Besides, though it's ok for current demand, building only ingress request and egress response may be a bit dumb and inflexible. I'm going to build a complete openapi spec, making the rust end more robust and general.

**Comment by n063h** [comment]
I'm adding some test cases on it. The maintainers can have a look at the updates now.

File: mosec/worker.py [hunk 8]
```diff
@@ -186,3 +189,30 @@ def forward(self, data: Any) -> Any:
...
+        cls, target: ParseTarget  # pylint: disable=unused-argument
+    ) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+        """Retrieve the JSON schema for the `forward` method of the class.
+
+        Args:
+            cls (type): The class object.
+            target (ParseTarget): The target variable to parse the schema for.
+
+        Returns:
+            Tuple[Dict[str, Any], Dict[str, Any]]: A tuple containing the schema
```
Comment (kemingy) [inline mosec/worker.py:204]: No need to add the type annotations here. The sphinx extension will parse it from the method annotations.

File: mosec/worker.py [hunk 9]
```diff
@@ -186,3 +189,30 @@ def forward(self, data: Any) -> Any:
...
+              and the component schemas.
+
+        The `get_forward_json_schema` method is a class method that returns the
+          JSON schema for the `forward` method of the given class `cls`.
+          It takes a `target` argument specifying the target to parse the schema for.
+        The returned value is a tuple containing the schema and the component schema.
```
Comment (kemingy) [inline mosec/worker.py:210]: The new paragraph requires an extra empty line. Check the generated doc.

File: mosec/utils/types.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    """
+    return parse_func_type(func, target, 0)
+
+
+def parse_cls_func_type(func, target: ParseTarget) -> type:
+    """Parse the input type of the target cls method.
+
+    The `class.func` remains the original func, so the index is 1.
+    """
+    return parse_func_type(func, target, 1)
```
Comment (kemingy) [inline mosec/utils/types.py:74]: Actually, you can use `inspect.ismethod` or `inspect.isfunction`.

File: src/main.rs [hunk 11]
```diff
@@ -67,13 +78,45 @@ async fn index(_: Request<Body>) -> Response<Body> {
...
+    )
+)]
+async fn openapi(_: Request<Body>, doc: openapi::OpenApi) -> Response<Body> {
+    let s = serde_json::to_string(&doc).unwrap_or("Openapi generation failed".to_string());
+    build_response(StatusCode::OK, Bytes::from(s))
+}
+
+#[utoipa::path(
+    post,
+    path = "/inference",
```
Comment (kemingy) [inline src/main.rs:109]: Is this configurable?

File: mosec/mixin/typed_worker.py [hunk 12]
```diff
@@ -59,26 +36,38 @@ class TypedMsgPackMixin(Worker):
     # pylint: disable=no-self-use
 
     resp_mime_type = "application/msgpack"
-    _input_type = None
+    input_typ: Optional[type] = None
```
Comment (kemingy) [inline mosec/mixin/typed_worker.py:39]: It should be a private attribute.

File: mosec/mixin/typed_worker.py [hunk 13]
```diff
@@ -59,26 +36,38 @@ class TypedMsgPackMixin(Worker):
...
         return msgspec.msgpack.encode(data)
+
+    @classmethod
+    def get_forward_json_schema(
+        cls, target: ParseTarget, ref_template: str
+    ) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+        """Get the JSON schema of the forward function."""
+        schema: Dict[str, Any]
+        comp_schema: Dict[str, Any]
+        (schema, comp_schema) = ({}, {})
```
Comment (kemingy) [inline mosec/mixin/typed_worker.py:65]:
```suggestion
        schema, comp_schema = {}, {}
```

File: mosec/worker.py [hunk 14]
```diff
@@ -186,3 +189,32 @@ def forward(self, data: Any) -> Any:
...
+        a `target` argument specifying the target to parse the schema for.
+
+        The returned value is a tuple containing the schema and the component schema.
+
+        Note:
+            Developer must implement this function to retrieve the JSON schema
+            to enable openapi spec.
+
+            The `MOSEC_REF_TEMPLATE` constant should be used as a reference template
+            according to openapi standards.
```
Comment (kemingy) [inline mosec/worker.py:218]: The inline code format is different from Markdown in reStructuredText.

File: src/apidoc.rs [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+use std::{collections::BTreeMap, str::FromStr};
+
+use serde::Deserialize;
+use utoipa::openapi::{
+    path::Operation, request_body::RequestBody, Components, OpenApi, PathItemType, RefOr, Response,
+    Responses, Schema,
+};
+
+#[derive(Deserialize, Default)]
+pub(crate) struct PythonApiDoc {
```
Comment (kemingy) [inline src/apidoc.rs:24]:
```suggestion
pub(crate) struct PythonAPIDoc {
```

Align with the naming convention in this project.

File: src/main.rs [hunk 16]
```diff
@@ -67,13 +86,70 @@ async fn index(_: Request<Body>) -> Response<Body> {
...
     let registry = REGISTRY.get().unwrap();
     encode(&mut encoded, registry).unwrap();
     build_response(StatusCode::OK, Bytes::from(encoded))
 }
 
+#[utoipa::path(
+    get,
+    path = "/openapi",
+    responses(
+        (status = StatusCode::OK, description = "Get Openapi Doc",body=String)
```
Comment (kemingy) [inline src/main.rs:111]:
```suggestion
        (status = StatusCode::OK, description = "Get OpenAPI doc",body=String)
```

File: src/apidoc.rs [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+        if let Some(mut responses) = python_api.responses {
+            let response = self.get_route_responses(&mut api, route, &PathItemType::Post);
+            response.responses.append(&mut responses);
+        };
+
+        MosecApiDoc { api }
+    }
+
+    pub fn move_path(&self, from: &str, to: &str) -> Self {
+        // move one path to another
```
Comment (kemingy) [inline src/apidoc.rs:109]: Both the function name and the docstring are not clear.

File: tests/test_service.py [hunk 18]
```diff
@@ -45,7 +45,7 @@ def mosec_service(request):
         shlex.split(f"python -u tests/{name}.py {args} --port {TEST_PORT}"),
     )
     assert wait_for_port_open(port=TEST_PORT), "service failed to start"
-    yield service
+    yield (service, args)
```
Comment (kemingy) [inline tests/test_service.py:48]: Will this change affect other test cases?

Comment (n063h) [reply tests/test_service.py:48]: No, Since our tests use HTTP requests to communicate, the popen instance `service` hasn't been used yet.

File: tests/openapi_service.py [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+    worker_mapping: Dict[str, Type[Worker]] = {
+        "TypedPreprocess": TypedPreprocess,
+        "UntypedPreprocess": UntypedPreprocess,
+        "TypedInference": TypedInference,
+        "UntypedInference": UntypedInference,
+    }
+
+    server = Server(endpoint="/v1/inference")
+    for w in sys.argv[1].split("/"):
+        server.append_worker(worker_mapping[w])
```
Comment (kemingy) [inline tests/openapi_service.py:80]: Seems this one doesn't activate the batch but the two `forward` methods use batch.

File: tests/test_service.py (hunk 18 above)
Comment (kemingy) [reply tests/test_service.py:48]: Better not do that. Since you can control the input to this fixture, you don't have to get the same input from this yield.

File: tests/openapi_service.py [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+class UntypedInference(TypedMsgPackMixin, Worker):
+    """Dummy batch inference."""
+
+    def forward(self, data):
+        return [len(buf) for buf in data]
+
+
+if __name__ == "__main__":
+    if len(sys.argv) < 2:
+        print("Please specify a shm to run: plasma or redis")
```
Comment (kemingy) [inline tests/openapi_service.py:68]: Unrelated.

**Review by kemingy — APPROVED** [review]
